### ISSUE TITLE
Grep filter doesn't work for systemd logs

### ISSUE THREAD

**Issue by jkschulz** [fluent/fluent-bit#1270]
## Bug Report

**Describe the bug**
Grep filter doesn't work for fields in systemd logs. 

**To Reproduce**
- Rubular link if applicable: https://rubular.com/r/j1QOxj0JOFBemK
- Steps to reproduce the problem:

== Config ==

```
    [INPUT]
        # https://docs.fluentbit.io/manual/input/systemd
        Name            systemd
        Alias           systemd.docker
        Tag             docker
        Path            /var/log/journal
        DB              /var/log/fluent-bit-k8s-node-journald-docker.db
        Systemd_Filter  _SYSTEMD_UNIT=docker.service

    [INPUT]
        # https://docs.fluentbit.io/manual/input/systemd
        Name            systemd
        Alias           systemd.kubelet
        Tag             kubelet
        Path            /var/log/journal
        DB              /var/log/fluent-bit-k8s-node-journald-kubelet.db
        Systemd_Filter  _SYSTEMD_UNIT=kubelet.service

    [INPUT]
        # https://docs.fluentbit.io/manual/input/systemd
        Name            systemd
        Alias           systemd.node-journal
        Tag             node-journal
        Path            /var/log/journal
        DB              /var/log/fluent-bit-k8s-node-journald.db

    [FILTER]
        # https://docs.fluentbit.io/manual/filter/grep
        Name     grep
        Alias    remove-duplicates
        Match    node-journal
        # Changing regex to "(docker|kubelet)\.service" or "docker|kubelet" has no effect.
        Exclude  _SYSTEMD_UNIT ^(docker|kubelet)\.service$

    [FILTER]
        # https://docs.fluentbit.io/manual/filter/stdout
        Name     stdout
        Match    *
```

== Result ==
stdout logs show `kubelet.service` and `docker.service` systemd logs being logged twice -- once with tag `kubelet`|`docker`, and once with tag `node-journal`
One example is in the rubular link. 

**Expected behavior**
systemd logs get tagged with "node-journal" only if _SYSTEMD_UNIT == !(docker.service OR kubelet.service).

more specifically:
- systemd logs with _SYSTEMD_UNIT=docker.service get tagged as "docker"
- systemd logs with _SYSTEMD_UNIT=kubelet.service get tagged as "kubelet"
- systemd logs with _SYSTEMD_UNIT=[anything that's not docker.service or kubelet.service] get tagged as "node-journal"

**Your Environment**

* Version used: fluent/fluent-bit:1.0.5-debug
* Environment name and version (e.g. Kubernetes? What version?): GKE (1.12.6-gke.10) cluster, with nodes running:
  * OS image: Container-Optimized OS from Google
  * Container runtime version: docker://17.3.2
  * kubelet version: v1.12.6-gke.10
* Filters and plugins: [input] systemd, [filter] grep, stdout

**Comment by edsiper** [comment]
Doing a quick look I think that the issue relies on the configuration:

- Every Systemd input instance that you configure works as an independent unit connected to Journald service, so if you don't specify specifically which Systemd_Units you care about, likely you will get the same records from multiple instances. One of your instances (number 2) do not specify the Unit:

```
    [INPUT]
        # https://docs.fluentbit.io/manual/input/systemd
        Name            systemd
        Alias           systemd.kubelet
        Tag             kubelet
        Path            /var/log/journal
        DB              /var/log/fluent-bit-k8s-node-journald-kubelet.db
        Systemd_Filter  _SYSTEMD_UNIT=kubelet.service

    [INPUT]
        # https://docs.fluentbit.io/manual/input/systemd
        Name            systemd
        Alias           systemd.node-journal
        Tag             node-journal
        Path            /var/log/journal
        DB              /var/log/fluent-bit-k8s-node-journald.db
```

My advice is: do not start two instances of Systemd input unless is necessary, if you need custom tagging please read next item.

- Tagging and Routing:  Systemd input plugin have a special feature to compose tags based in the Systemd_Unit name they come from. Consider the following example:

```
        Name         systemd
        Alias           systemd.node-journal
        Tag             node-journal.*
        Path            /var/log/journal
        DB              /var/log/fluent-bit-k8s-node-journald.db
```

the wildcard or star will be replaced with the Systemd Unit name, so messages coming from Docker will be tagged as node-journal.docker, while others coming from kubelet will get node-journal.kubelet. That's the common way to take advantage of tagging in Systemd, that's whats most of our users do.

- Filter Grep: grep is not designed to remove duplicates, just to keep or exclude certain records based on patters and it works in a loop iterating each record. 

At this point you don't need this _filter grep_ since the root cause of the problem is the Input setup.

**Comment by edsiper** [comment]
and... please use the latest v1.1.2 version (v1.2 will be released shortly)

**Comment by jkschulz** [comment]
The systemd logs we're trying to process have ~12 (we can't know exactly how many or what they are) different _SYSTEMD_UNIT values. We want to be able to tag ones that are "kubelet.service" as "kubelet", "docker.service" as "docker", and everything else as "node-journal".

The node-journal input doesn't specify a unit in order to capture _all_ systemd logs, and then the subsequent filter (should, I think) omit the "kubelet" and "docker" logs, since those were already captured by the other inputs. 

Is there a configuration you would suggest to accomplish this goal? I don't think Systemd_Filter accepts a regex, so the only other option I could see was to have a systemd input for _every_ potential _SYSTEMD_UNIT value, with most tagging "node-journal".

**Comment by edsiper** [comment]
There is no support to exclude certain units in Systemd, take a look at this:

https://npm.pkg.github.com/systemd/systemd/issues/2720

So what I can think as a workaround (and quite experimental honestly) is to take advante of our new Stream Processor engine and try something like this:

__fluent-bit.conf__

```
[SERVICE]
    Flush   1
    Streams_File  stream_processor.conf

[INPUT]
    Name    systemd
    Alias   systemd

[OUTPUT]
    Name    stdout
    Match   kubelet

[OUTPUT]
    Name    stdout
    Match   docker

[OUTPUT]
    Name    stdout
    Match   node-journal
```

__stream_processor.conf__

```
[STREAM_TASK]
    Name  unit_kubelet
    Exec  CREATE STREAM sp_kubelet WITH(tag='kubelet') AS SELECT * FROM STREAM:systemd WHERE _SYSTEMD_UNIT = 'kubelet.service';

[STREAM_TASK]
    Name  unit_docker
    Exec  CREATE STREAM sp_docker WITH(tag='docker') AS SELECT * FROM STREAM:systemd WHERE _SYSTEMD_UNIT = 'docker.service';

[STREAM_TASK]
    Name  all_others
    Exec  CREATE STREAM sp_others WITH(tag='node-journal') AS SELECT * FROM STREAM:systemd WHERE NOT _SYSTEMD_UNIT = 'kubelet.service' AND NOT _SYSTEMD_UNIT = 'docker.service';
```

the only thing is that this requires change 02f73b66e94a747e7c6828e3ffc8ad62b535ebe8 that I've just submitted and is not part of an official release. But you can try:

1. build Fluent Bit from GIT master
2. Use the following dev/test Docker image: edsiper/fluent-bit-1.2-next:2

**Comment by qingling128** [comment]
@edsiper - That looks promising. Do we know when the next Fluent Bit release will be?

**Comment by edsiper** [comment]
@qingling128 as soon as we finish fixing this issue: https://github.com/fluent/fluent-bit/issues/1278 . It's almost done, just waiting for users-feedback.

**Comment by qingling128** [comment]
@edsiper - Thanks! Looking forward to the release.

**Comment by qingling128** [comment]
@edsiper - For the version, we need to wait for a `1.2` release, right? I just noticed a release (https://github.com/fluent/fluent-bit/releases/tag/v1.1.3) has gone out, but it did not seem to contain the commit we needed (https://github.com/fluent/fluent-bit/commit/02f73b66e94a747e7c6828e3ffc8ad62b535ebe8).

**Comment by edsiper** [comment]
@qingling128 yes, wait until v1.2 release

**Comment by edsiper** [comment]
1.2 being released now, thanks.

**Comment by qingling128** [comment]
Thanks @edsiper!

**Comment by jkschulz** [comment]
I still can't get this to work using v1.2.0



**fluent-bit.conf**
```
    [SERVICE]
        # https://docs.fluentbit.io/manual/service
        Flush         1
        Streams_File  stream_processor.conf
        Log_Level     warn
        Daemon        off
        Parsers_File  parsers.conf
        HTTP_Server   On
        HTTP_Listen   0.0.0.0
        HTTP_Port     2020
        # https://docs.fluentbit.io/manual/configuration/buffering
        storage.path               /var/log/fluent-bit-buffers/
        storage.sync               normal
        storage.checksum           off
        storage.backlog.mem_limit  10M

    @INCLUDE input-containers.conf
    @INCLUDE input-systemd.conf
    @INCLUDE filter-kubernetes.conf
    @INCLUDE output-fluentd.conf
```
**input-systemd.conf**
 ```
   [INPUT]
        Name  systemd
        Alias systemd

    [OUTPUT]
        Name  stdout
        Match kubelet

    [OUTPUT]
        Name  stdout
        Match docker

    [OUTPUT]
        Name  stdout
        Match node-journal
```
**stream_processor.conf**
```
    [STREAM_TASK]
      Name systemd_kubelet
      Exec CREATE STREAM sp_kubelet WITH(tag='kubelet') AS SELECT * FROM STREAM:systemd WHERE _SYSTEMD_UNIT = 'kubelet.service';

    [STREAM_TASK]
      Name systemd_docker
      Exec CREATE STREAM sp_docker WITH(tag='docker') AS SELECT * FROM STREAM:systemd WHERE _SYSTEMD_UNIT = 'docker.service';

    [STREAM_TASK]
      Name systemd_node_journal
      Exec CREATE STREAM sp_others WITH(tag='node-journal') AS SELECT * FROM STREAM:systemd WHERE NOT _SYSTEMD_UNIT = 'kubelet.service' AND NOT _SYSTEMD_UNIT = 'docker.service';
```

Here's one of the outputs where the log is being tagged as `node-journal` when it has `"_SYSTEMD_UNIT"=>"kubelet.service"`:
```
[25] node-journal: [1562718375.926973000, {"_BOOT_ID"=>"a7d24d9381af4f1fa9aaa0c85ffc9ede", "_MACHINE_ID"=>"fe8e39e60017d6984d66ea29c0754e4e", "PRIORITY"=>"6", "SYSLOG_FACILITY"=>"3", "_UID"=>"0", "_GID"=>"0", "_SYSTEMD_SLICE"=>"system.slice", "_CAP_EFFECTIVE"=>"3fffffffff", "_TRANSPORT"=>"stdout", "_HOSTNAME"=>"gke-jschulz-gke-default-pool-227f0ea3-7fn8", "SYSLOG_IDENTIFIER"=>"kubelet", "_COMM"=>"kubelet", "_EXE"=>"/home/kubernetes/bin/kubelet", "_CMDLINE"=>"/home/kubernetes/bin/kubelet --v=2 --cloud-provider=gce --experimental-check-node-capabilities-before-mount=true --allow-privileged=true --experimental-mounter-path=/home/kubernetes/containerized_mounter/mounter --cert-dir=/var/lib/kubelet/pki/ --cni-bin-dir=/home/kubernetes/bin --kubeconfig=/var/lib/kubelet/kubeconfig --experimental-kernel-memcg-notification=true --max-pods=110 --network-plugin=kubenet --node-labels=beta.kubernetes.io/fluentd-ds-ready=true,cloud.google.com/gke-nodepool=default-pool,cloud.google.com/gke-os-distribution=cos --volume-plugin-dir=/home/kubernetes/flexvolume --bootstrap-kubeconfig=/var/lib/kubelet/bootstrap-kubeconfig --node-status-max-images=25 --registry-qps=10 --registry-burst=20 --config /home/kubernetes/kubelet-config.yaml", "_SYSTEMD_CGROUP"=>"/system.slice/kubelet.service", "_SYSTEMD_UNIT"=>"kubelet.service", "_PID"=>"1373", "_SYSTEMD_INVOCATION_ID"=>"caaf7730ee9d42f199278cece8e5fce0", "MESSAGE"=>"I0710 00:26:15.925944    1373 reconciler.go:301] Volume detached for volume "ssl-certs" (UniqueName: "kubernetes.io/host-path/0b66eb32-a2a7-11e9-9583-42010a8e0089-ssl-certs") on node "gke-jschulz-gke-default-pool-227f0ea3-7fn8" DevicePath """, "logging.googleapis.com/local_resource_id"=>"k8s_node.gke-jschulz-gke-default-pool-227f0ea3-7fn8"}]
```

**Comment by nokute78** [comment]
@jkschulz Would you try this workaround?
(not use `(` and `)` as regex )
```
[FILTER]
    Name     grep
    Match    *
    Exclude  _SYSTEMD_UNIT docker\.service|kubelet\.service
```

I tested this config, and it seems to work.

```python
[INPUT]
    Name  dummy
    Tag   test
    dummy {"_SYSTEMD_UNIT":"kubelet.service"}

[INPUT]
    Name  dummy
    Tag   test
    dummy {"_SYSTEMD_UNIT":"docker.service"}

[INPUT]
    Name  dummy
    Tag   test
    dummy {"_SYSTEMD_UNIT":"OTHER"}

[FILTER]
    Name     grep
    Match    *
    Exclude  _SYSTEMD_UNIT docker\.service|kubelet\.service

[OUTPUT]
    Name   stdout
    Match  *
```


kubelet.service and docker.service are filtered.
```
[taka@localhost 1270]$ ../bin/fluent-bit -c a.conf 
Fluent Bit v1.3.0
Copyright (C) Treasure Data

[2019/07/10 18:02:19] [ info] [storage] initializing...
[2019/07/10 18:02:19] [ info] [storage] in-memory
[2019/07/10 18:02:19] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2019/07/10 18:02:19] [ info] [engine] started (pid=27781)
[0] test: [1562749340.001121661, {"_SYSTEMD_UNIT"=>"OTHER"}]
[1] test: [1562749341.000171564, {"_SYSTEMD_UNIT"=>"OTHER"}]
[2] test: [1562749342.001179215, {"_SYSTEMD_UNIT"=>"OTHER"}]
[3] test: [1562749343.001454105, {"_SYSTEMD_UNIT"=>"OTHER"}]
```

**Comment by nokute78** [comment]
@edsiper I wrote a issue #1433 .
I think it is a root cause.

**Comment by jkschulz** [comment]
@nokute78 Ah, that worked! Thanks! 
Does your new issue (#1433) also address the problem with the stream processing?

**Comment by nokute78** [comment]
@jkschulz Great! 

> Does your new issue (#1433) also address the problem with the stream processing?

No. It is to fix regex. 
If it is addressed , your regex https://github.com/fluent/fluent-bit/issues/1270#issue-431712014 should work.
```
        Exclude  _SYSTEMD_UNIT ^(docker|kubelet)\.service$
```

**Comment by edsiper** [comment]
@jkschulz what's the remaining issue with the stream processor ?

**Comment by jkschulz** [comment]
@edsiper The stream processing config in the [comment above](https://github.com/fluent/fluent-bit/issues/1270#issuecomment-509866380) is separate from the regex, but doesn't isolate `kubelet` or `docker` logs from those tagged `node-journal`, even with this bit of config:
```
 [STREAM_TASK]
      Name systemd_node_journal
      Exec CREATE STREAM sp_others WITH(tag='node-journal') AS SELECT * FROM STREAM:systemd WHERE NOT _SYSTEMD_UNIT = 'kubelet.service' AND NOT _SYSTEMD_UNIT = 'docker.service';
```

**Comment by edsiper** [comment]
@jkschulz in order to let the records stop after they hit the stream processor you can set "Routable false", e.g:

```
[INPUT]
    Name   systemd
    Alias     systemd
    Routable false
```

records with "routable" turned off (or false), are discarded right away. This is new and landed in v1.2:

https://fluentbit.io/announcements/v1.2.0/

note: please use v1.2.1

**Comment by jkschulz** [comment]
@edsiper I don't want the records to be discarded, I just want records without `_SYSTEMD_UNIT` == `kubelet.service`|`docker.service` to be tagged as `node-journal`. 

I feel like the query you gave me was exactly what I was looking for:
```
Exec CREATE STREAM sp_others WITH(tag='node-journal') AS SELECT * FROM STREAM:systemd WHERE NOT _SYSTEMD_UNIT = 'kubelet.service' AND NOT _SYSTEMD_UNIT = 'docker.service';
```
but when I tried it in my config, it was still tagging _everything_ as `node-journal` (there's a sample log demonstrating this in one of my earlier comments). Can you take another look at the config and see if I'm missing anything?

**Comment by edsiper** [comment]
@jkschulz I don't see any issue in the config, would you please re-paste here the actual config ? (fresh copy from your config files)
